### PR TITLE
several improvements for authentication

### PR DESCRIPTION
**Pull request by maxtaco**
- allow UID or username, but both not required
- allow email, in addition, or instead of, the above
- testing

### CODE REVIEW

**Comment by malgorithms** [comment]
@maxtaco  grab me when you want to walk through it

**Comment by maxtaco** [comment]
Grabbing you!!!

File: src/base.iced [hunk 1]
```diff
@@ -244,15 +249,46 @@ class Base
...
+    if json?.body?.key?.uid
+      if (a = json?.body?.key?.uid) isnt (b = @user.local.uid)
+        return new Error "Wrong local uid: got '#{a}' but wanted '#{b}'"
+      else
+        has_user_id = true
+    else if @_v_require_uid()
+      return new Error "no uid given, but was was required"
+
+    if (v = @user.local.emails)? and (e = json?.body?.key?.email)?
+      if e.toLowerCase() in v
```
Comment (maxtaco) [inline src/base.iced]: run `toLowerCase` on all emails in `v` too; or just not `toLowerCase` at all.

File: test/files/auth.iced [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  email = new_email()
+  test_auth T, {
+    gen_hook : (garg) ->
+      garg.user.local.email = email
+    verify_hook : (varg, carg) ->
+      carg.user.local.emails = [ "x" + email ]
+    err_hook : (err) ->
+      T.assert err?, "errored out on bad email"
+  }, cb
+
```
Comment (maxtaco) [inline test/files/auth.iced]: check casing if we're leaving casing in.

**Pull request merged**